src/shiftedCompositeNormL2.jl

```diff
@@ -0,0 +1,228 @@
+  b = similar(ψ.b)
+  ψ.c!(b,xk)
+  A = similar(ψ.A)
+  ψ.J!(A,xk)
+  ShiftedCompositeNormL2(ψ.h.lambda, ψ.c!, ψ.J!, A, b)
+end
+ 
+shifted(
+  ψ::CompositeNormL2{R, V0, V1, V2, V3},
+  xk::AbstractVector{R}
+) where {R <: Real, V0 <: Function, V1 <: Function, V2 <: AbstractMatrix{R}, V3 <: AbstractVector{R}} = begin
+  b = similar(ψ.b)
+  ψ.c!(b,xk)
+  A = copy(ψ.A)
+  ψ.J!(A,xk)
+  ShiftedCompositeNormL2(ψ.h.lambda, ψ.c!, ψ.J!, A, b)
+end
+
+fun_name(ψ::ShiftedCompositeNormL2) = "shifted L2 norm"
+fun_expr(ψ::ShiftedCompositeNormL2) = "t ↦ ‖c(xk) + J(xk)t‖₂"
+fun_params(ψ::ShiftedCompositeNormL2) = "c(xk) = $(ψ.b)\n" * " "^14 * "J(xk) = $(ψ.A)\n"
+
+function prox!(
+  y::AbstractVector{R},
+  ψ::ShiftedCompositeNormL2{R, V0, V1, V2, V3},
+  q::AbstractVector{R},
+  σ::R;
+  max_iter = 10000,
+  max_time = 180.0
+) where {R <: Real, V0 <: Function,V1 <:Function,V2 <: AbstractMatrix{R}, V3 <: AbstractVector{R}}
+
+  start_time = time()
+  θ = R(0.8)
+  γ = R(0.0)
+  full_row_rank = true
+  # Initialize Aᵧ
+  ψ.Aᵧ.rows .= [ψ.A.rows;collect(eltype(ψ.A.rows),1:ψ.A.m)] 
+  ψ.Aᵧ.cols .= [ψ.A.cols;collect(eltype(ψ.A.cols),ψ.A.n+1:ψ.A.n + ψ.A.m)]
+  ψ.Aᵧ.vals .= [ψ.A.vals;zeros(eltype(ψ.A.vals),ψ.A.m)]
+
+  mul!(ψ.g, ψ.A, q)
+  ψ.g .+= ψ.b
+
+  spmat = qrm_spmat_init(ψ.Aᵧ; sym=false) # TODO: preallocate this
+  spfct = qrm_spfct_init(spmat) # TODO: preallocate this
+  qrm_set(spfct, "qrm_keeph", 0)
+
+  # Check interior convergence
+  qrm_analyse!(spmat, spfct; transp='t')
+  qrm_factorize!(spmat, spfct, transp='t')
+
+  qrm_solve!(spfct, ψ.g, ψ.p, transp='t')
+  qrm_solve!(spfct, ψ.p, ψ.sol, transp='n')
+
+  # 1 Step of iterative refinement
+  ψ.res .= ψ.g
+
+  mul!(ψ.dp, ψ.Aᵧ', ψ.sol)
+  mul!(ψ.dsol, ψ.Aᵧ, ψ.dp)
+
+  ψ.res .-= ψ.dsol
+  if norm(ψ.res) > eps(R)^0.75
+    qrm_solve!(spfct, ψ.res, ψ.dp, transp='t')
+    qrm_solve!(spfct, ψ.dp, ψ.dsol, transp='n')
+    ψ.sol .+= ψ.dsol
+  end  
+
+  ψ.sol .*= -1
+
+  # Check full row rankness of J(x) by inspecting diagonal of R
+  R1 = qrm_spfct_get_r(spfct)
+  cp = qrm_spfct_get_cp(spfct)
+  rp = qrm_spfct_get_rp(spfct)
+  rows_R = Int[] #TODO : preallocate this
+
+  for i = 1 : size(R1, 1)
+    if nnz(R1[rp[i],cp[:]]) != 0
+      push!(rows_R, i)
+    end
+  end
+  for i = 1 : size(R1, 2)
+    if abs(R1[rp[rows_R[i]], cp[i]]) < eps(R)^0.75
+      full_row_rank = false
+      break
+    end
+  end
+
+  if !full_row_rank 
+
+    γ = eps(R)
+    
+    qrm_update!(spmat,[ψ.A.vals; fill(eltype(ψ.A.vals)(sqrt(γ)),ψ.A.m)])
+    qrm_factorize!(spmat, spfct, transp='t')
+
+    qrm_solve!(spfct, ψ.g, ψ.p, transp='t')
+    qrm_solve!(spfct, ψ.p, ψ.sol, transp='n')
+    qrm_solve!(spfct, ψ.sol, ψ.p, transp='t')
+    _iterative_refinement!(spfct, ψ)
+
+    ψ.sol .*= -1
+
+    if norm(ψ.sol) ≤ σ*ψ.h.lambda + eps(R) # We are in the hard-case, we consider ψ.sol is a good approximation of the least-norm solution
+      mul!(y, ψ.A', ψ.sol)
+      y .+= q
```

Comment:
Here, something happened with a - sign. I think it's because the right-hand side had the wrong sign in the system with $R^T R$ that you solved above. Please explain in comments.